ssl/statem/statem_clnt.c

```diff
@@ -1562,11 +1563,22 @@ MSG_PROCESS_RETURN tls_process_server_certificate(SSL *s, PACKET *pkt)
 
     exp_idx = ssl_cipher_get_cert_index(s->s3->tmp.new_cipher);
     if (exp_idx >= 0 && i != exp_idx) {
-        x = NULL;
-        al = SSL_AD_ILLEGAL_PARAMETER;
-        SSLerr(SSL_F_TLS_PROCESS_SERVER_CERTIFICATE,
-               SSL_R_WRONG_CERTIFICATE_TYPE);
-        goto f_err;
+        if (exp_idx == SSL_PKEY_GOST_EC) {
+            if ((i != SSL_PKEY_GOST12_512) && (i != SSL_PKEY_GOST12_256)
+                && (i != SSL_PKEY_GOST01)) {
+                x = NULL;
+                al = SSL_AD_ILLEGAL_PARAMETER;
+                SSLerr(SSL_F_TLS_PROCESS_SERVER_CERTIFICATE,
+                       SSL_R_WRONG_CERTIFICATE_TYPE);
+                goto f_err;
+            }
+        } else {
+            x = NULL;
+            al = SSL_AD_ILLEGAL_PARAMETER;
+            SSLerr(SSL_F_TLS_PROCESS_SERVER_CERTIFICATE,
+                   SSL_R_WRONG_CERTIFICATE_TYPE);
+            goto f_err;
+        }
```

Comment:
Can we collapse this into a single "if"? I don't like that we have the same code repeated twice for the inner "if" and the "else".